timely/src/dataflow/operators/filter.rs

```diff
@@ -24,15 +25,47 @@ pub trait Filter<D: Data> {
 
 impl<G: Scope, D: Data> Filter<D> for Stream<G, D> {
     fn filter<P: FnMut(&D)->bool+'static>(&self, mut predicate: P) -> Stream<G, D> {
-        let mut vector = Vec::new();
-        self.unary(Pipeline, "Filter", move |_,_| move |input, output| {
-            input.for_each(|time, data| {
-                data.swap(&mut vector);
-                vector.retain(|x| predicate(x));
-                if !vector.is_empty() {
-                    output.session(&time).give_vec(&mut vector);
+        let mut builder = OperatorBuilder::new("Filter".to_owned(), self.scope());
+
+        let mut input = builder.new_input_filter(self, Pipeline, vec![], move |data, buffer| {
+            match data {
+                RefOrMut::Ref(data) => {
+                    buffer.extend(data.into_iter().filter(|x| predicate(*x)).cloned())
```

Comment:
Fairly unrelated, but it'd be nice if operators could pass on references instead of requiring cloning, that could remove a lot of fairly redundant clones

Reply:
This might do what you suggested: https://github.com/TimelyDataflow/timely-dataflow/pull/471